exchanges/poloniex/poloniex_wrapper.go

```diff
@@ -491,259 +648,563 @@ func (e *Exchange) SubmitOrder(ctx context.Context, s *order.Submit) (*order.Sub
+					resp.Status[co.OrderID] = cancellationStatus
+				} else {
+					resp.Status[co.ClientOrderID] = cancellationStatus
+				}
+			}
 		}
-		resp.Status[cancelledOrders[i].OrderID] = cancelledOrders[i].State + " " + cancelledOrders[i].Message
 	}
 	return resp, nil
 }
 
 // CancelAllOrders cancels all orders associated with a currency pair
-func (e *Exchange) CancelAllOrders(ctx context.Context, _ *order.Cancel) (order.CancelAllResponse, error) {
+func (e *Exchange) CancelAllOrders(ctx context.Context, cancelOrd *order.Cancel) (order.CancelAllResponse, error) {
 	cancelAllOrdersResponse := order.CancelAllResponse{
 		Status: make(map[string]string),
 	}
-	openOrders, err := e.GetOpenOrdersForAllCurrencies(ctx)
-	if err != nil {
-		return cancelAllOrdersResponse, err
+	if cancelOrd == nil {
+		return cancelAllOrdersResponse, common.ErrNilPointer
 	}
-
-	for key := range openOrders.Data {
-		for i := range openOrders.Data[key] {
-			err = e.CancelExistingOrder(ctx, openOrders.Data[key][i].OrderNumber)
+	var pairs currency.Pairs
+	if !cancelOrd.Pair.IsEmpty() {
+		fPair, err := e.FormatExchangeCurrency(cancelOrd.Pair, cancelOrd.AssetType)
+		if err != nil {
+			return cancelAllOrdersResponse, err
+		}
+		pairs = append(pairs, fPair)
+	}
+	switch cancelOrd.AssetType {
+	case asset.Spot:
+		switch cancelOrd.Type {
+		case order.TrailingStop, order.TrailingStopLimit, order.StopLimit, order.Stop:
+			var orderTypes []OrderType
+			if cancelOrd.Type != order.UnknownType {
+				orderTypes = append(orderTypes, OrderType(cancelOrd.Type))
+			}
+			resp, err := e.CancelSmartOrders(ctx, pairs, nil, orderTypes)
 			if err != nil {
-				id := strconv.FormatInt(openOrders.Data[key][i].OrderNumber, 10)
-				cancelAllOrdersResponse.Status[id] = err.Error()
+				return cancelAllOrdersResponse, err
+			}
+			for _, co := range resp {
+				cancelAllOrdersResponse.Status[co.OrderID] = co.State
+			}
+		default:
+			if e.Websocket.IsConnected() && e.Websocket.CanUseAuthenticatedEndpoints() && e.Websocket.CanUseAuthenticatedWebsocketForWrapper() {
+				wsResponse, err := e.WsCancelTradeOrders(ctx, pairs.Strings(), []AccountType{AccountType(cancelOrd.AssetType)})
+				if err != nil {
+					return cancelAllOrdersResponse, err
+				}
+				for _, wco := range wsResponse {
+					if wco.Code == 0 || wco.Code == 200 {
+						cancelAllOrdersResponse.Status[strconv.FormatUint(wco.OrderID, 10)] = wco.State
+					} else {
+						cancelAllOrdersResponse.Status[strconv.FormatUint(wco.OrderID, 10)] = "failed"
+					}
+				}
+			} else {
+				resp, err := e.CancelTradeOrders(ctx, pairs.Strings(), []AccountType{AccountType(cancelOrd.AssetType)})
+				if err != nil {
+					return cancelAllOrdersResponse, err
+				}
+				for _, co := range resp {
+					if co.Code == 0 || co.Code == 200 {
+						cancelAllOrdersResponse.Status[co.OrderID] = co.State
+					} else {
+						cancelAllOrdersResponse.Status[co.OrderID] = "Failed"
+					}
+				}
 			}
 		}
+	case asset.Futures:
```

Comment:
has no code coverage

If it does, please stop using `require.NoError` and use `assert.NoError` for multi-case tests